### PR TITLE
BUG: Merge with str/StringDtype keys and multiindex

### LINKED ISSUE #43734: BUG: Merge with EA and MultiIndex
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd

df1 = pd.DataFrame(data={("lvl0","lvl1-a"): ["1","2","3"], ("lvl0","lvl1-b"): ["4","5","6"]}, dtype=str)

df2 = pd.DataFrame(data={("lvl0","lvl1-a"): ["1","2","3"], ("lvl0","lvl1-c"): ["7","8","9"]}, dtype=pd.StringDtype())

result = pd.merge(left=df1, right=df2, on=[("lvl0","lvl1-a")])
```

### Issue Description

I am trying to make a merge between two data frames with a `MultiIndex` for the columns. The first data frame has an object data type (`str` in my example) and the second an extension array data type (`pd.StringDtype()` in my example). In this case, in an internal method, an `assign` method is used but unfortunately, the `name` of the column is not a string but a tuple...

### Expected Behavior

I would have expected the merge to happen as it is with two `str` dtype data frames or two `pd.StirngDtype()` dtype data frames.

```
    lvl0              
  lvl1-a lvl1-b lvl1-c
0      1      4      7
1      2      5      8
2      3      6      9
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 73c68257545b5f8530b7044f56647bd2db92e2ba
python           : 3.9.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.0-8-amd64
Version          : #1 SMP Debian 5.10.46-4 (2021-08-03)
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : fr_FR.UTF-8
LOCALE           : fr_FR.UTF-8
pandas           : 1.3.3
numpy            : 1.21.2
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 57.4.0
Cython           : 0.29.24
pytest           : 6.2.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 3.0.1
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.1
sqlalchemy       : 1.4.25
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by benoit9126**
- [x] closes #43734
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/reshape/merge/test_merge.py [hunk 1]
```diff
@@ -1885,6 +1885,17 @@ def test_dtype_on_merged_different(self, change, join_type, left, right):
         )
         tm.assert_series_equal(result, expected)
 
+        # GH 43734 Avoid the use of `assign` with multiindex
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py:1888]: pls make a new test

File: pandas/tests/reshape/merge/test_merge.py [hunk 2]
```diff
@@ -1885,6 +1885,17 @@ def test_dtype_on_merged_different(self, change, join_type, left, right):
...
+        right.columns = MultiIndex.from_tuples([("lvl0", x) for x in right.columns])
+        left.columns = MultiIndex.from_tuples([("lvl0", x) for x in left.columns])
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py:1890]: this doesn't appear to actually be testing the OP, e.g. that's where the dtype is str vs StringDtype(). can you parameterize and show the dataframe construction.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/merge/test_merge.py (hunk 1 above)
Comment (benoit9126) [reply pandas/tests/reshape/merge/test_merge.py]: I created a new test and I parametrized the Python string data type (as you proposed) and the join type (as it was in the test I originally inserted this fragment of code)

File: doc/source/whatsnew/v1.4.0.rst [hunk 3]
```diff
@@ -502,6 +502,7 @@ Reshaping
 - Bug in :func:`concat` of ``bool`` and ``boolean`` dtypes resulting in ``object`` dtype instead of ``boolean`` dtype (:issue:`42800`)
 - Bug in :func:`crosstab` when inputs are are categorical Series, there are categories that are not present in one or both of the Series, and ``margins=True``. Previously the margin value for missing categories was ``NaN``. It is now correctly reported as 0 (:issue:`43505`)
 - Bug in :func:`concat` would fail when the ``objs`` argument all had the same index and the ``keys`` argument contained duplicates (:issue:`43595`)
+- Fixed bug in :meth:`merge` with multi-index as column index for the ``on`` argument returning an error when assigning a column internally (:issue:`43734`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.4.0.rst:505]: We use func for merge

Comment (phofl) [reply doc/source/whatsnew/v1.4.0.rst:505]: also :class:`MultiIndex`

File: pandas/core/reshape/merge.py [hunk 4]
```diff
@@ -1280,10 +1280,12 @@ def _maybe_coerce_merge_keys(self) -> None:
             # incompatible dtypes. See GH 16900.
             if name in self.left.columns:
                 typ = lk.categories.dtype if lk_is_cat else object
-                self.left = self.left.assign(**{name: self.left[name].astype(typ)})
+                self.left = self.left.copy()
```
Comment (phofl) [inline pandas/core/reshape/merge.py:1283]: Why is the copy necessary?

File: pandas/tests/reshape/merge/test_merge.py [hunk 5]
```diff
@@ -2598,3 +2598,31 @@ def test_merge_outer_with_NaN(dtype):
         dtype=dtype,
     )
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("string_dtype", tm.STRING_DTYPES)
+def test_merge_ea_with_string(join_type, string_dtype):
```
Comment (phofl) [inline pandas/tests/reshape/merge/test_merge.py:2604]: Your test does not cover the copy, works even without.

Comment (benoit9126) [reply pandas/tests/reshape/merge/test_merge.py:2604]: It looks difficult to do. The `self.left` (and `self.right`) are internal objects of the `_MergeOperation` class... 

I can ensure that `df1` and `df2` dtypes have not changed after the call to `merge` in my test but, even without the `copy`, it continues to work. What would you propose ?

Comment (phofl) [reply pandas/tests/reshape/merge/test_merge.py:2604]: Switching df1 and df2 in the merge call

File: pandas/tests/reshape/merge/test_merge.py [hunk 6]
```diff
@@ -2598,3 +2598,37 @@ def test_merge_outer_with_NaN(dtype):
...
+        dtype=string_dtype,
+    )
+    merged = merge(left=df1, right=df2, on=[("lvl0", "lvl1-a")], how=join_type)
+
+    # No change in df1 and df2 types
+    assert (df1.dtypes == pd.StringDtype()).all()
+    assert (df2.dtypes == np.dtype("O")).all()
+
+    # Check the expected types for the merged data frame
+    result = merged.dtypes.sort_index()
```
Comment (phofl) [inline pandas/tests/reshape/merge/test_merge.py:2627]: Don't use sort, just define expected correctly

File: pandas/tests/reshape/merge/test_merge.py [hunk 7]
```diff
@@ -2598,3 +2598,37 @@ def test_merge_outer_with_NaN(dtype):
...
+        data={
+            ("lvl0", "lvl1-a"): ["1", "2", "3", pd.NA, "5"],
+            ("lvl0", "lvl1-c"): ["7", "8", "9", pd.NA, "11"],
+        },
+        dtype=string_dtype,
+    )
+    merged = merge(left=df1, right=df2, on=[("lvl0", "lvl1-a")], how=join_type)
+
+    # No change in df1 and df2 types
+    assert (df1.dtypes == pd.StringDtype()).all()
```
Comment (phofl) [inline pandas/tests/reshape/merge/test_merge.py:2623]: Check whole DataFrames

File: pandas/tests/reshape/merge/test_merge.py (hunk 6 above)
Comment (benoit9126) [reply pandas/tests/reshape/merge/test_merge.py]: I picked this way of doing in the class `TestMergeCategorical` (in the same file). Done.

File: pandas/tests/reshape/merge/test_merge.py (hunk 7 above)
Comment (benoit9126) [reply pandas/tests/reshape/merge/test_merge.py]: Done

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Pull request merged**